### ISSUE TITLE
Suggestion: support inline previews of tab completions

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PSReadLine#741]
<sup>Note: The suggested feature may have to be constrained to certain types of completions, and it is debatable whether it should be the default behavior or opt-in via a new option, but overall I think it would greatly streamline the tab-completion experience.</sup>

Imagine if PSReadLine showed you a _preview_ of what the token you're currently typing _would_ expand to _if_ you pressed <kbd>Tab</kbd> at that moment; e.g.:

![image](https://user-images.githubusercontent.com/588825/43689396-f5f8ef0e-98c7-11e8-9717-a3b99fdf63e4.png)

That is, having typed parameter-name prefix `-Fo`, PSReadLine would show you, via the differently colored - "rce" part, that  _if_ you pressed <kbd>Tab</kbd> _at that moment_ you'd end up with `-Force`.

If the preview doesn't indicate the desired parameter, simply _keep typing_ until the desired one is suggested; e.g., additionally type an `l` to make the preview show "lowSymlink"  to inform you that pressing <kbd>Tab</kbd> then would complete to `-FollowSymlink`.

This makes for a very convenient experience _if you know what exactly what completion you're looking for_ : just keep typing - without interruption - until you see what you want, then press <kbd>Tab</kbd>.

A challenge is how to handle transitioning to _reviewing all possible completions_ at a given point, given that pressing <kbd>Tab</kbd> in this scenario would have to result in _instant_ completion (whereas currently you can use it to _cycle through the candidate completions_).

 * If technically feasible, pressing <kbd>Tab</kbd> _right after_ a completion would still be aware of what prefix was actually typed and would still allow cycling through the possible completions.

* Alternatively / complementarily, a separate key chord such as <kbd>Ctrl +  Tab</kbd>could be used to show all possibly completions in a _menu_, via built-in function `MenuComplete`.

---

As an aside: If someone prefers the current tab-completion behavior in principle, but would like to show a _menu_ of possible completions in case of ambiguity, they can add the following to your `$PROFILE`:

```powershell
Set-PSReadlineKeyHandler -Key Tab -Function MenuComplete
```

Environment data
----------------

```powershell
PS version: 6.1.0-preview.4
PSReadline version: 2.0.0-beta2
os: 10.0.16299.15 (WinBuild.160101.0800)
PS file version: 6.1.0
```

**Comment by felixfbecker** [comment]
Duplicate of https://github.com/PowerShell/PSReadLine/issues/687

**Comment by mklement0** [comment]
Thanks, @felixfbecker  - hadn't seen yours. Certainly closely related, but your suggestion is about completions from the _history_, whereas this one is about the standard, syntax-based tab completions.

Both behaviors are useful, but I'm unclear on if and how they could be rolled into a single feature: when would you choose which completion source?

**Comment by TBBle** [comment]
It appears that https://github.com/PowerShell/CompletionPredictor delivers this feature via Prediction. The default keys are different because it's Prediction (being fed from Completion) rather than Completion, but that does also resolve the question in https://github.com/PowerShell/PSReadLine/issues/741#issuecomment-490915663, because you get suggestions from [both history and plugins by default](https://devblogs.microsoft.com/powershell/psreadline-2-2-6-enables-predictive-intellisense-by-default/#predictions-are-enabled-by-default), and in the list view (but not inline, see #1945) it tells you where the prediction came from.

**Comment by mklement0** [comment]
Thanks, @TBBle - it looks  this now ships with PSReadLine 2.1+  (for _history_-based completions, still not available for _syntax_-based completions, which is what this issue is about), but must be activated explicitly, e.g. with `Set-PSReadLineOption -PredictionSource History`; see [this blog post](https://devblogs.microsoft.com/powershell/announcing-psreadline-2-1-with-predictive-intellisense/) for details.

The default completion style is `InlineView`, but that can be changed with `Set-PSReadLineOption -PredictionViewStyle ListView`

To see all relevant PSReadLine Options:

```powershell
Get-PSReadLineOption | oss | Select-String Prediction
```

In PSReadLine 2.2+, _custom_ completers can be defined as _plugins_ (requiring activation via either `Set-PSReadLineOption -PredictionSource HistoryAndPlugin` or `Set-PSReadLineOption -PredictionSource Plugin`) and  https://github.com/PowerShell/CompletionPredictor can serve as a sample implementation; its read-me points to [How to create a command-line predictor](https://docs.microsoft.com/en-us/powershell/scripting/dev-cross-plat/create-cmdline-predictor).

**Comment by daxian-dbw** [comment]
The ask of this issue can be achieved by using the `CompletionPredictor` module along with PSReadLine 2.2+. Especially, the predictive intellisense feature is enabled by default in [v2.2.6 release](https://github.com/PowerShell/PSReadLine/releases/tag/v2.2.6) of PSReadLine.

Here is a GIF to show the asked behavior (using `Windows` edit mode and `HistoryAndPlugin` as the PredictionSource):

![CompletionPredictor](https://user-images.githubusercontent.com/127450/176494972-30bf0988-2f72-4231-9edf-248dc5273ecd.gif)

Given this, I will mark the issue as fixed.

**Comment by TBBle** [comment]
The plugin https://github.com/PowerShell/CompletionPredictor adds (If I understand correctly) "_syntax_-based completions". In PSReadLine 2.2.6, `Set-PSReadLineOption -PredictionSource HistoryAndPlugin` is the default, so installing and importing that module should be sufficient to resolve this ticket.

(And I can see that's happening in parallel as I comment)

**Comment by mklement0** [comment]
Good to know and glad to hear it - thank you.

**Comment by mklement0** [comment]
One more thing, because it may not be obvious (should someone in the future stumble upon this thread): plugins (completer modules) must be imported _explicitly_ into every session (via `$PROFILE`) - PSReadLine, at last as of v2.2.6, does _not_ try to discover and load them automatically (for performance reasons, I presume).
